src/Microsoft.AspNetCore.OData/Extensions/Endpoint/ODataEndpointLinkGenerator.cs

```diff
@@ -0,0 +1,135 @@
+
+                    bool canGenerateDirectLink = routePrefix == null || routePrefix.IndexOf('{') == -1;
+                    if (!canGenerateDirectLink)
+                    {
+                        routePrefix = BindPrefixTemplate(routePrefix, values, ambientValues);
+                    }
+                    string link = CombinePathSegments(routePrefix, odataPath);
+                    link = UriEncode(link);
+                    return link;
+                }
+            }
+
+            return _innerGenerator.GetPathByAddress(httpContext, address, values, ambientValues, pathBase, fragment, options);
+        }
+
+        /// <inheritdoc/>
+        public override string GetPathByAddress<TAddress>(TAddress address, RouteValueDictionary values, PathString pathBase = default, FragmentString fragment = default, LinkOptions options = null)
+        {
+            return _innerGenerator.GetPathByAddress<TAddress>(address, values, pathBase, fragment, options);
+        }
+
+        /// <inheritdoc/>
+        public override string GetUriByAddress<TAddress>(HttpContext httpContext, TAddress address, RouteValueDictionary values, RouteValueDictionary ambientValues = null, string scheme = null, HostString? host = null, PathString? pathBase = null, FragmentString fragment = default, LinkOptions options = null)
+        {
+            return _innerGenerator.GetUriByAddress(httpContext, address, values, ambientValues, scheme, host, pathBase, fragment, options);
+        }
+
+        /// <inheritdoc/>
+        public override string GetUriByAddress<TAddress>(TAddress address, RouteValueDictionary values, string scheme, HostString host, PathString pathBase = default, FragmentString fragment = default, LinkOptions options = null)
+        {
+            return _innerGenerator.GetUriByAddress(address, values, scheme, host, pathBase, fragment, options);
+        }
+
+        // Noted: simple workaround to bind the value for the prefix template
+        // Shold replace it using the standard ASP.NET Core way later.
+        private static string BindPrefixTemplate(string prefix, RouteValueDictionary values, RouteValueDictionary ambientValues)
+        {
+            IList<string> templates = new List<string>();
+            int startIndex = 0;
+            while (true)
+            {
+                int start = prefix.IndexOf('{', startIndex);
+                int end = prefix.IndexOf('}', start + 1); // start == -1 is ok
+                if (start == -1 || end == -1)
+                {
+                    break;
+                }
+
+                startIndex = end + 1;
+
+                string subStr = prefix.Substring(start, end - start + 1);
+                templates.Add(subStr);
+            }
+
+            foreach (var item in templates)
+            {
+                string variable = item.Substring(1, item.Length - 2); // remove { and }
+
```

Comment:
Why substring again? We should add to templates after striping the braces already.

Reply:
This loop can easily be eliminated by stripping the braces before adding them to the collection

Reply:
We only replace the string within "{" and "}". and keep the other as it. for example:

prefix:  "data{data}" and we have {"data": 1} in route value, the return of this method should be "data1", not "11". See more from my added test cases.

---
In reply to: [376002864](https://github.com/OData/WebApi/pull/2035#discussion_r376002864) [](ancestors = 376002864)